provider/semgrep/index.ts

```diff
@@ -0,0 +1,128 @@
+    Annotation,
+    AnnotationsParams,
+    AnnotationsResult,
+    ItemsParams,
+    ItemsResult,
+    MentionsParams,
+    MentionsResult,
+    MetaParams,
+    MetaResult,
+    Provider,
+} from '@openctx/provider'
+import { XMLBuilder } from 'fast-xml-parser'
+
+import { type Finding, type Findings, urlfor } from './api.js'
+import API from './client.js'
+
+export type Settings = {
+    deployment: string // Semgrep deployment slug
+    token: string // Semgrep app token
+    repo?: string // Semgrep scan repository
+}
+
+function parseUri(link: string): { deployment: string; finding: number } | null {
+    const url = new URL(link)
+    if (url.hostname.includes('semgrep.dev')) {
+        try {
+            const [d, f] = url.pathname.split('/orgs/')[1].split('/findings/')
+            return { deployment: d, finding: parseInt(f) }
+        } catch (err) {
+            return null
+        }
+    }
+    return null
+}
+
+function aiPrompt(finding: Finding): string {
+    const xml = new XMLBuilder({ format: true })
+    const info = {
+        rule: {
+            name: (finding.rule_name || finding.rule.name) ?? null,
+            message: (finding.rule_message || finding.rule.message) ?? null,
+        },
+        triage: {
+            state: finding.triage_state ?? null,
+            reason: finding.triage_reason ?? null,
+            comment: finding.triage_comment ?? null,
+        },
+        assistant: {
+            autofix: {
+                fixCode: finding.assistant?.autofix?.fix_code ?? null,
+                explanation: finding.assistant?.autofix?.explanation ?? null,
+            },
+            autotriage: {
+                reason: finding.assistant?.autotriage?.reason ?? null,
+            },
+        },
+    }
+    return xml.build(info)
+}
+
+const semgrep: Provider = {
+    meta(params: MetaParams, settings: Settings): MetaResult {
+        return { name: 'Semgrep', annotations: {}, mentions: {} }
+    },
+
+    items(params: ItemsParams, settings: Settings): ItemsResult {
+        if (params.mention?.data?.finding === undefined) return []
+        const finding: Finding = params.mention.data.finding as Finding
+        return !finding
+            ? []
+            : [
+                  {
+                      title: finding.rule_name,
+                      ai: { content: aiPrompt(finding) },
+                      ui: { hover: { text: finding.rule_name } },
+                      url: urlfor(settings.deployment, finding.repository.name, finding.id),
+                  },
+              ]
+    },
+
+    async mentions(params: MentionsParams, settings: Settings): Promise<MentionsResult> {
+        if (params.query) {
+            const q: { deployment: string; finding: number } | null = parseUri(params.query)
+            if (q) {
+                const client: API = new API({
+                    repo: settings.repo,
+                    token: settings.token,
+                    deployment: q.deployment,
```

Comment:
here you use the deployment that was parsed out of the params.query (rather than from settings). But then in items you use settings.deployment for the URL. I assume that is incorrect in items? Instead you should be parsing out the deployment from the mention.uri.

Is the only point of settings.deployment for annotations? If so I would consider making it optional, or if removing annotations support remove deployments from the config.

Reply:
Thanks, that's a good spot. Yeah, `settings.deployment` is more useful for annotations. Nevertheless, I'll make it optional and use it as a fallback in case `mentions.uri` fails to populate for some reason.

Reply:
Addressed in https://github.com/sourcegraph/openctx/pull/105/commits/d2cd580b69a3d18a0a4ae7b156f79269b800377f & https://github.com/sourcegraph/openctx/pull/105/commits/70961707541974ebc922e10e70ebbc55395f4d68